README.md
# TJSP Scraper

A simple scraper for the detail page of a single litigation from the Justice Court of the State of São Paulo.

## Install Clojure and Leiningen 

Clojure: https://clojure.org/guides/getting_started
Leiningen: https://leiningen.org/#install

## Usage

Just run the clojure_project-0.1.0-SNAPSHOT-standalone.jar file using the number of a judicial process from São Paulo.

Ex:

java -jar target/tjsp_scraper-0.1.0-SNAPSHOT-standalone.jar 10673490420178260100

java -jar target/tjsp_scraper-0.1.0-SNAPSHOT-standalone.jar 1097007-73.2017.8.26.0100

## License

/*
 * ----------------------------------------------------------------------------
 * "THE BEER-WARE LICENSE" (Revision 42):
 * <daniel.fernandesrey@gmail.com> wrote this file.  As long as you retain this notice you
 * can do whatever you want with this stuff. If we meet some day, and you think
 * this stuff is worth it, you can buy me a beer in return.
 * ----------------------------------------------------------------------------
 */
=======





CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-08-02
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-08-02
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clojure_project/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clojure_project/compare/0.1.0...0.1.1


test/tjsp_scraper/scraper_test.clj
(ns tjsp-scraper.scraper_test
  (:require [clojure.test :refer :all]
            [tjsp-scraper.scraper :refer :all]
            [clojure.string :as strs]))

(deftest scraper-test

  (testing "get-process-params"
    (let [numeracao-unica "11277690920168260100"
          params (get-process-params numeracao-unica)
          query-params (get params :query-params)
          numero-digito-ano-unificado (get query-params :numeroDigitoAnoUnificado)
          foro-numero-unificado (get query-params :forumNumeroUnificado)
          ]
      (is (instance? clojure.lang.PersistentArrayMap params))
      (is (= numero-digito-ano-unificado (subs numeracao-unica 0 13)))
      (is (= foro-numero-unificado (subs numeracao-unica 16)))
      )
    )
  (testing "extract-process-info"
    (let [numeracao-unica "11277690920168260100"
          process-info (extract-process-info numeracao-unica)
          ]
      (is (contains? process-info :dados))
      (is (contains? process-info :partes))
      (is (contains? process-info :movimentacoes))
      (is (instance? clojure.lang.PersistentHashMap process-info))
      )
    )

  )



src/clojure_project/core.clj
(ns clojure-project.core
  (:require [org.httpkit.client :as http]
            [net.cgrand.enlive-html :as html]
            [clojure.string :as strs])
  (:gen-class))

(def url "http://esaj.tjsp.jus.br/cpopg/search.do")

(defn get-process-params [numeracao_unica]
  "Return a hash-map representing the get params to request it's information."
  (let [numero (strs/replace numeracao_unica #"\.|-" "")
        numeroDigitoAnoUnificado (subs numero 0 13)
        foroNumeroUnificado (subs numero 16)
        ]

    hash-map {:query-params {
                             :conversationId                         ""
                             :dadosConsulta.localPesquisa.cdLocal    "-1"
                             :cbPesquisa                             "NUMPROC"
                             :dadosConsulta.tipoNuProcesso           "UNIFICADO"
                             :numeroDigitoAnoUnificado               numeroDigitoAnoUnificado
                             :forumNumeroUnificado                   foroNumeroUnificado
                             :dadosConsulta.valorConsultaNuUnificado numero
                             :dadosConsulta.valorConsulta            ""
                             }
              :insecure?    true
              }))

(defn get-dom
  "Get the dom from a page as a html/html-snippet element."
  [url numeracao_unica]
  (html/html-snippet
    (:body @(http/get url (get-process-params numeracao_unica)))))

(defn get-movs-info
  "Get the tbody the containing the content of all movs"
  [content]
  (html/select content [:tbody#tabelaTodasMovimentacoes])
  )

(defn get-trs
  "Find all trs elements from the mov"
  [movs-info]
  (html/select movs-info [:tr])
  )

(defn get-movs-content [tr-text]
  "Extract the text information from a 'movimentacao' field from the process."
  (strs/trim (get (strs/split tr-text #"[0-9]{1,2}/[0-9]{1,2}/[0-9]{1,4} ") 1))
  )

(defn get-movs-date [tr-text]
  "Find the mov date"
  (first (re-seq #"[0-9]{1,2}/[0-9]{1,2}/[0-9]{1,4}" tr-text))
  )

(defn not-blank? [str]
  "Check if a string is not blank"
  (not (strs/blank? str))
  )

(defn get-info

  "Receives a vector, a start position and a end position and return a subvector of the vector"

  ([lista start] (subvec lista start))
  ([lista start end] (subvec lista start end))

  )


(defn clean [texto]
  "Clean the text of a mov"
  (strs/replace texto #"\t|\n" ""))

(defn get-header
  "Get the header of the page"
  ([lista positions] (get-header lista positions [] 0))

  ([lista positions result x]

    ;Recursive get a list of string representing all of the information from the header of the process
   (if (not= x (- (count positions) 1) )
     (let [cont (+ x 1)
           local-result (conj result (get-info lista (nth positions x) (nth positions cont) ) )
           ]
       (recur lista positions local-result cont)
       )

     (conj result (get-info lista (last positions) ))
     ))
  )

(defn get-process-dados [dom]
  "Get the 'dados' field from a judicial process."
  (let [table (last (html/select dom [:table.secaoFormBody]))
        texts-list (strs/split (strs/trim (clean (html/text table) ) ) #"\s{2,}")
        index-elements (keep-indexed (fn [index item] [index (strs/ends-with? item ":" )]) texts-list)
        keys (filter (fn [x] (true? (last x) ) ) index-elements)
        positions (map (fn [x] (first x) ) keys)
        ]

    (get-header texts-list positions)
    )
  )


(defn get-partes-info [tr]
  "Get the information related to the 'partes' field of a judicial process."
  (let [parte-text (clean (html/text tr))
        partes-advogados-list (filterv not-blank? (strs/split parte-text #"  "))

        ]
    partes-advogados-list
    )
  )


(defn get-partes [dom]
  "Extract raw information about the 'partes' of a process and return it as list of lists."
  (let [tabela (html/select dom [:table#tableTodasPartes])
        trs (get-trs tabela)
        partes (map get-partes-info trs)
        ]
    partes
    )
  )

(defn get-movimentacoes [dom]
  "Extract the information related to the 'movimentaçõeses' field of a process."
  (let [movs-content (get-movs-info dom)
        trs-content (get-trs movs-content)
        text-content (map html/text trs-content)
        cleaned-text (map clean text-content)
        movs-date (map get-movs-date cleaned-text)
        movs-content (map get-movs-content cleaned-text)
        lista-movimentacoes (map vector movs-date movs-content)
        ]
    lista-movimentacoes
    )
  )

(defn -main [numeracao_unica]
  "Main function.
  Returns a hash-map containing 3 keys and it's values.
  The keys are 'dados' 'partes' and 'movimentações.'"
  (let [dom (get-dom url numeracao_unica)
        dados (get-process-dados dom)
        partes (get-partes dom)
        lista-movimentacoes (get-movimentacoes dom)
        dados-processo (hash-map :movimentacoes lista-movimentacoes
                                 :partes partes
                                 :dados dados)
        ]
    (println dados-processo)

    )
  )


src/tjsp_scraper/scraper.clj
(ns tjsp-scraper.scraper
  (:require [org.httpkit.client :as http]
            [net.cgrand.enlive-html :as html]
            [clojure.string :as strs]
            [clojure.data.json :as json])
  (:gen-class))

(def url "http://esaj.tjsp.jus.br/cpopg/search.do")

(defn get-process-params [numeracao-unica]
  "Return a hash-map representing the get params to request it's information."

  (if (nil? numeracao-unica)
    {}
    (let [numero (strs/replace numeracao-unica #"\.|-" "")
          numero-digito-ano-unificado (subs numero 0 13)
          foro-numero-unificado (subs numero 16)
          ]

      {:query-params {
                      :conversationId                         ""
                      :dadosConsulta.localPesquisa.cdLocal    "-1"
                      :cbPesquisa                             "NUMPROC"
                      :dadosConsulta.tipoNuProcesso           "UNIFICADO"
                      :numeroDigitoAnoUnificado               numero-digito-ano-unificado
                      :forumNumeroUnificado                   foro-numero-unificado
                      :dadosConsulta.valorConsultaNuUnificado numero
                      :dadosConsulta.valorConsulta            ""
                      }
       :insecure?    true
       }))
  )



(defn get-body
  [url params]
  (:body @(http/get url params)))

(defn get-dom
  "Get the dom from a page as a html/html-snippet element."

  ([url] (get-dom url nil))
  ([url numeracao-unica]
   (html/html-snippet
     (get-body url (get-process-params numeracao-unica)))
    )

  )

(defn get-movs-info
  "Get the tbody the containing the content of all movs"
  [content]
  (html/select content [:tbody#tabelaTodasMovimentacoes])
  )

(defn get-trs
  "Find all trs elements from the mov"
  [movs-info]
  (html/select movs-info [:tr])
  )

(defn get-movs-content [tr-text]
  "Extract the text information from a 'movimentacao' field from the process."
  (strs/trim (get (strs/split tr-text #"[0-9]{1,2}/[0-9]{1,2}/[0-9]{1,4} ") 1))
  )

(defn get-movs-date [tr-text]
  "Find the mov date"
  (first (re-seq #"[0-9]{1,2}/[0-9]{1,2}/[0-9]{1,4}" tr-text))
  )

(defn not-blank? [str]
  "Check if a string is not blank"
  (not (strs/blank? str))
  )

(defn get-info
  "Receives a vector, a start position and a end position and return a subvector of the vector"
  ([lista start]
   (get-info lista start (count lista)))
  ([lista start end]
   (subvec lista start end))
  )

(defn vec->map
  "Convert a vector with two positions to a map"
  [line-vec]
  (let [key (first line-vec)
        value (strs/join " " (subvec line-vec 1))
        ]
    {key value}
    )
  )

(defn get-header
  "Get the header of the page"
  [lista positions]
  (let [line-vec (get-info lista (first positions) (last positions))]
    (vec->map line-vec)
    )
  )

(defn clean [texto]
  "Clean a text"
  (strs/replace texto #"\t|\n" ""))

(defn get-process-dados [dom]
  "Get the 'dados' field from a judicial process."
  (let [table (last (html/select dom [:table.secaoFormBody]))
        texts-list (strs/split (strs/trim (clean (html/text table))) #"\s{2,}")
        chaves (filter (fn [x] (strs/ends-with? x ":")) texts-list)
        positions (map (fn[chaves] (.indexOf texts-list chaves)) chaves)
        processs-dados (mapv #(get-header texts-list %1) (partition 2 1 positions))
        info->map (comp vec->map get-info)
        ]

    (->> (info->map texts-list (last positions))
         (conj processs-dados)
         (reduce into {})
         )

    )
  )


(defn get-partes-info [tr]
  "Get the information related to the 'partes' field of a judicial process."
  (let [parte-text (clean (html/text tr))
        partes-advogados-list (filterv not-blank? (strs/split parte-text #"  "))

        ]
    partes-advogados-list
    )
  )

(defn get-partes [dom]
  "Extract raw information about the 'partes' of a process and return it as list of lists."
  (let [tabela (html/select dom [:table#tableTodasPartes])
        trs (get-trs tabela)
        partes (map get-partes-info trs)
        ]
    partes
    )
  )

(defn get-movimentacoes [dom]
  "Extract the information related to the 'movimentações' field of a process."
  (let [movs-content (get-movs-info dom)
        trs-content (get-trs movs-content)
        text-content (map html/text trs-content)
        cleaned-text (map clean text-content)
        movs-date (map get-movs-date cleaned-text)
        movs-content (map get-movs-content cleaned-text)
        lista-movimentacoes (map vector movs-date movs-content)
        ]
    lista-movimentacoes
    )
  )

(defn get-file-name
  "Get the file name to write the contento of the process to"
  [numeracao_unica]
  (->> (strs/replace numeracao_unica #"\.|-" "")
       (format "%s.json")
       )

  )

(defn save-dados-processo
  "Save the info from the processo to a json file."
  [dados-processo nome-arquivo]
  (spit nome-arquivo (json/write-str dados-processo))
  )

(defn extract-process-info
  "Returns a hash-map containing 3 keys and it's values.
  The keys are 'dados' 'partes' and 'movimentações'."

  ([numeracao-unica] (extract-process-info numeracao-unica false))
  ([numeracao-unica write-to-disk]
   (let [dom (get-dom url numeracao-unica)
         dados (get-process-dados dom)
         partes (get-partes dom)
         lista-movimentacoes (get-movimentacoes dom)
         dados-processo (hash-map :movimentacoes lista-movimentacoes
                                  :partes partes
                                  :dados dados
                                  :numeracao-unica numeracao-unica)
         ]
     (if write-to-disk
       (save-dados-processo dados-processo (get-file-name numeracao-unica))
       )
     dados-processo
     ))
  )

(defn -main
  "Main function."
  ([numeracao_unica] (-main numeracao_unica false))
  ([numeracao_unica write-to-disk]
   (println (extract-process-info numeracao_unica write-to-disk))
    )

  )

